File: frontend/src/types/commune.ts
/**
 * Types for commune data and scoring system
 * 
 * Scores are now dynamic - loaded from the backend based on available datasets.
 */

/** Geographic coordinates */
export interface Coordinates {
  lat: number;
  lng: number;
}

/** Base commune information */
export interface CommuneBase {
  codeInsee: string;
  nom: string;
  departement: string;
  departementCode: string;
  region: string;
  regionCode: string;
  population: number;
  coordinates: Coordinates;
}

/**
 * Dynamic scores - keys are filter IDs from the manifest.
 * This allows adding new filters without changing the type definition.
 */
export type CommuneScores = Record<string, number>;

/** Commune with computed scores */
export interface CommuneWithScore extends CommuneBase {
  scores: CommuneScores;
  /** Weighted global score (0-100) */
  scoreGlobal: number;
}

/** Commune summary for list views */
export interface CommuneSummary {
  codeInsee: string;
  nom: string;
  departement: string;
  scoreGlobal: number;
  coordinates: Coordinates;
}

/** GeoJSON feature for a commune */
export interface CommuneGeoJSONFeature {
  type: 'Feature';
  properties: {
    codeInsee: string;
    nom: string;
    scoreGlobal: number;
    population: number;
    scores?: CommuneScores;
  };
  geometry: {
    type: 'Polygon' | 'MultiPolygon';
    coordinates: number[][][] | number[][][][];
  };
}

/** GeoJSON collection for communes */
export interface CommunesGeoJSON {
  type: 'FeatureCollection';
  features: CommuneGeoJSONFeature[];
}

/** Score level for UI display */
export type ScoreLevel = 'excellent' | 'good' | 'average' | 'poor' | 'bad';

/** Helper to get score level from numeric score */
export function getScoreLevel(score: number): ScoreLevel {
  if (score >= 80) return 'excellent';
  if (score >= 60) return 'good';
  if (score >= 40) return 'average';
  if (score >= 20) return 'poor';
  return 'bad';
}

/**
 * Filter category metadata from the API.
 * Loaded dynamically from the backend based on available datasets.
 */
export interface FilterCategory {
  id: string;
  name: string;
  description: string;
  icon: string;
  unit: string | undefined;
  weightDefault: number;
  datasetId: string;
}

/**
 * Data status from the API.
 * Shows what data is available and if the app is ready.
 */
export interface DataStatus {
  communesAvailable: boolean;
  communesFile: string | null;
  datasets: Array<{
    id: string;
    name: string;
    enabled: boolean;
    fileExists: boolean;
    filterCount: number;
  }>;
  filterCount: number;
  ready: boolean;
}

File: backend/app/models/commune.py
"""
Commune data models.

Scores are now dynamic - keys are filter IDs from the manifest.
"""


from pydantic import BaseModel, ConfigDict, Field


class Coordinates(BaseModel):
    """Geographic coordinates."""

    lat: float = Field(..., ge=-90, le=90, description="Latitude")
    lng: float = Field(..., ge=-180, le=180, description="Longitude")


class CommuneBase(BaseModel):
    """Base commune information."""

    code_insee: str = Field(..., min_length=5, max_length=5, description="Code INSEE de la commune")
    nom: str = Field(..., min_length=1, description="Nom de la commune")
    departement: str = Field(..., description="Nom du département")
    departement_code: str = Field(..., min_length=2, max_length=3, description="Code du département")
    region: str = Field(..., description="Nom de la région")
    region_code: str = Field(..., min_length=2, max_length=3, description="Code de la région")
    population: int = Field(..., ge=0, description="Population de la commune")
    coordinates: Coordinates = Field(..., description="Coordonnées géographiques du centroïde")


# Dynamic scores - keys are filter IDs from manifest
CommuneScores = dict[str, float]


class CommuneSummary(BaseModel):
    """Summary view of a commune for lists."""

    code_insee: str = Field(..., description="Code INSEE")
    nom: str = Field(..., description="Nom de la commune")
    departement: str = Field(..., description="Nom du département")
    score_global: float = Field(..., ge=0, le=100, description="Score global (0-100)")
    coordinates: Coordinates = Field(..., description="Coordonnées")


class CommuneDetail(CommuneBase):
    """Detailed commune information with all scores."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "code_insee": "75056",
                "nom": "Paris",
                "departement": "Paris",
                "departement_code": "75",
                "region": "Île-de-France",
                "region_code": "11",
                "population": 2161000,
                "coordinates": {"lat": 48.8566, "lng": 2.3522},
                "scores": {
                    "climat_chaleur_extreme": 75.0,
                    "climat_incendie": 80.0,
                },
                "score_global": 77.5,
            }
        }
    )

    scores: dict[str, float] = Field(
        default_factory=dict,
        description="Scores par catégorie (filter_id -> score 0-100)"
    )
    score_global: float = Field(..., ge=0, le=100, description="Score global pondéré (0-100)")
